File: src/telecom/matcher/telecom-subscriber-matcher.ts
import type { CellTower } from '../../types/tower.js';
import { loadConfig, type ParsedEnvConfig } from '../../config/env.js';
import { getLogger } from '../../utils/logger.js';
import type { SubscriberMatch, SubscriberMatcher } from '../../pipeline/subscriber-matcher.js';
import type { SubscriberRepository } from '../repositories/subscriber-repository.js';

const log = getLogger();

/**
 * TelecomSimSubscriberMatcher — modules 03/04 drop-in.
 *
 * Implements the pipeline's SubscriberMatcher contract against the simulated
 * subscriber repository. For every tower the cell-site resolver returned, it
 * fetches the subscribers currently attached to that cell (chunked, via the
 * repository) and returns their real MSISDNs.
 *
 * Because it speaks the same interface the future C-DOT matcher will use,
 * swapping in the real subscriber database requires no pipeline changes —
 * only `registerSubscriberMatcher(new CdotSubscriberMatcher(...))`.
 */
export class TelecomSimSubscriberMatcher implements SubscriberMatcher {
  readonly name = 'telecom-sim';

  constructor(
    private readonly repo: SubscriberRepository,
    private readonly cfg: ParsedEnvConfig = loadConfig(),
  ) {}

  async matchSubscribers(
    towers: readonly CellTower[],
    ctx: { alertId: string; capIdentifier: string },
  ): Promise<SubscriberMatch[]> {
    const { alertId, capIdentifier } = ctx;
    const start = performance.now();

    const cellIds = towers.map((t) => t.cellId).filter((c) => c);
    const uniqueCellIds = Array.from(new Set(cellIds));

    const rows = await this.repo.findByCellIds(uniqueCellIds, { limit: this.cfg.TOWER_MATCH_LIMIT });
    const byCell = new Map<string, string[]>();
    for (const row of rows) {
      const list = byCell.get(row.cellId) ?? [];
      list.push(row.msisdn);
      byCell.set(row.cellId, list);
    }

    const matches: SubscriberMatch[] = towers.map((tower) => ({
      towerId: tower.id,
      msisdns: byCell.get(tower.cellId) ?? [],
    }));

    const matched = matches.reduce((n, m) => n + m.msisdns.length, 0);
    const elapsedMs = performance.now() - start;
    log.info({ alertId, capIdentifier, towers: towers.length, matched, elapsedMs }, 'subscriber.match.completed');
    return matches;
  }
}

File: src/main/java/com/turant/subscriber/TelecomSubscriberMatcher.java
package com.turant.subscriber;

import com.turant.types.tower.CellTower;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.stereotype.Service;

import java.util.*;
import java.util.concurrent.CompletableFuture;
import java.util.stream.Collectors;

/**
 * Telecom subscriber matcher - modules 03/04 drop-in.
 * 
 * Implements the pipeline's SubscriberMatcher contract against the simulated
 * subscriber repository. For every tower the cell-site resolver returned, it
 * fetches the subscribers currently attached to that cell (chunked, via the
 * repository) and returns their real MSISDNs.
 * 
 * Migrated from TypeScript telecom/matcher/telecom-subscriber-matcher.ts
 */
@Service
public class TelecomSubscriberMatcher implements SubscriberMatcher {
    
    private static final Logger logger = LoggerFactory.getLogger(TelecomSubscriberMatcher.class);
    
    private final SubscriberRepository repository;
    
    @Value("${tower.match-limit:100000}")
    private int towerMatchLimit;
    
    public TelecomSubscriberMatcher(SubscriberRepository repository) {
        this.repository = repository;
    }
    
    @Override
    public String getName() {
        return "telecom-sim";
    }
    
    @Override
    public CompletableFuture<List<SubscriberMatch>> matchSubscribers(
            List<CellTower> towers,
            MatchContext ctx) {
        
        long startTime = System.currentTimeMillis();
        
        // Extract unique cell IDs
        List<String> cellIds = towers.stream()
            .map(CellTower::cellId)
            .filter(Objects::nonNull)
            .distinct()
            .collect(Collectors.toList());
        
        logger.info("Matching subscribers: alertId={}, towers={}, uniqueCells={}", 
            ctx.alertId(), towers.size(), cellIds.size());
        
        // Fetch subscribers by cell IDs
        return repository.findByCellIds(
            cellIds, 
            new SubscriberRepository.FindByCellIdsOptions().setLimit(towerMatchLimit)
        ).thenApply(rows -> {
            // Group subscribers by cell ID
            Map<String, List<String>> byCellId = new HashMap<>();
            for (SubscriberRepository.SubscriberRow row : rows) {
                byCellId.computeIfAbsent(row.cellId(), k -> new ArrayList<>())
                    .add(row.msisdn());
            }
            
            // Create matches for each tower
            List<SubscriberMatch> matches = towers.stream()
                .map(tower -> new SubscriberMatch(
                    tower.id(),
                    byCellId.getOrDefault(tower.cellId(), Collections.emptyList())
                ))
                .collect(Collectors.toList());
            
            int totalMatched = matches.stream()
                .mapToInt(m -> m.msisdns().size())
                .sum();
            
            long elapsedMs = System.currentTimeMillis() - startTime;
            
            logger.info("Subscriber matching completed: alertId={}, capIdentifier={}, towers={}, matched={}, elapsedMs={}", 
                ctx.alertId(), ctx.capIdentifier(), towers.size(), totalMatched, elapsedMs);
            
            return matches;
        });
    }
}
